src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+#include <algorithm>  // std::transform
+#include <cmath>
+#include <fmt/format.h>
+#include <sstream>
+#include <type_traits>
+
+#include <json.hpp>
+
+#include <bbp/sonata/node_sets.h>
+
+namespace bbp {
+namespace sonata {
+
+namespace {
+using json = nlohmann::json;
+
+const size_t MAX_COMPOUND_RECURSION = 10;
+
+template <typename T>
+std::string toString(const std::string& key, const std::vector<T>& values) {
+    return fmt::format(R"("{}": [{}])", key, fmt::join(values, ", "));
+}
+
+template <>
+std::string toString(const std::string& key, const std::vector<std::string>& values) {
+    // strings need to be wrapped in quotes
+    return fmt::format(R"("{}": ["{}"])", key, fmt::join(values, "\", \""));
+}
+
+template <typename T>
+class NodeSetBasicRule: public NodeSetRule
+{
+  public:
+    NodeSetBasicRule(std::string attribute, std::vector<T>&& values)
+        : attribute_(std::move(attribute))
+        , values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& v : values_) {
+            ret = ret | np.matchAttributeValues(attribute_, v);
```

Comment:
maybe add a comment inside the function to remind the dev the float comparisons are not allowed.

Reply:
I added it to the NodeSets docstring